mesh/mesh.hpp

```diff
@@ -978,22 +994,41 @@ class Mesh
 
        The IntegrationRule used with GetFaceGeometricFactors needs to remain
        valid until the internally stored FaceGeometricFactors objects are
-       destroyed (by either calling Mesh::DeleteGeometricFactors or the Mesh
-       destructor). */
-   const FaceGeometricFactors* GetFaceGeometricFactors(const IntegrationRule& ir,
-                                                       const int flags,
-                                                       FaceType type,
-                                                       MemoryType d_mt = MemoryType::DEFAULT);
+       destroyed (by either calling Mesh::DeleteGeometricFactors(),
+       Mesh::NodesUpdated(), or the Mesh destructor).
+
+       If the device MemoryType parameter @a d_mt is specified, then the
+       returned object will use that type unless it was previously allocated
+       with a different type.
+
+       The returned pointer points to an internal object that may be invalidated
+       by mesh operations such as refinement, vertex/node movement, etc. Since
+       not all such modifications can be tracked by the Mesh class (e.g. when
+       using the pointer returned by GetNodes() to change the nodes) one needs
+       to account for such changes by calling the method NodesUpdated() which,
+       in particular, will call DeleteGeometricFactors(). */
+   const FaceGeometricFactors* GetFaceGeometricFactors(
+      const IntegrationRule& ir,
+      const int flags,
+      FaceType type,
+      MemoryType d_mt = MemoryType::DEFAULT);
 
    /// Destroy all GeometricFactors stored by the Mesh.
    /** This method can be used to force recomputation of the GeometricFactors,
-       for example, after the mesh nodes are modified externally. */
+       for example, after the mesh nodes are modified externally.
+
+       @note In general, the preferred method for resetting the GeometricFactors
+       should be to call NodesUpdated(). */
    void DeleteGeometricFactors();
 
-   /// @brief This function should be called after the mesh node coordinates
-   /// have changed, e.g. after the mesh has moved.
-   /** It updates internal quantities derived from the node coordinates, such
-       as the GeometricFactors. */
+   /** @brief This function should be called after the mesh node coordinates
+       have been updated externally, e.g. by modifying the internal nodal
+       GridFunction returned by GetNodes(). */
+   /** It updates/deletes internal quantities derived from the node coordinates,
+       such as the (Face)GeometricFactors.
+
+       @note Unlike the similarly named protected method UpdateNodes() this
+       method does not modify the nodes. */
    void NodesUpdated() { DeleteGeometricFactors(); }
```

Comment:
If `DeleteGeometricFactors()` and `NodesUpdated` do exactly the same thing and one is "preferred", why do both exist?

regarding "updates/deletes internal quantities", does it actually ever update the geometric factors, or does it only delete them? If so, when is something updated, rather than deleted?

Reply:
I believe `NodesUpdated` is intended to be user-facing, and not necessarily tied to the implementation details of the geometric factors (e.g. in the future, `NodesUpdated` could do something more/different than just `DeleteGeometricFactors`). The idea is that users shouldn't need to know anything about the internal geometric factors (or even their existence), but just that they need to register changes to the nodes grid function with this function (because unfortunately we cannot track those changes since the user can access the raw nodes data).

`DeleteGeometricFactors` on the other hand is low-level, and just clears the cached geometric factors. I'm not sure if we are considering making this protected in the future…

To answer your second question, it only deletes the geometric factors, which are then updated/constructed lazily when they are next used.

Reply:
Just FYI: we use `DeleteGeometricFactors` in our application to trigger re-computation of geo-facts.

Reply:
To add on we use it to free GPU memory when we know they are going to need to be recomputed later but we want the memory back now

Reply:
```suggestion
   /** It deletes internal quantities derived from the node coordinates,
       such as the (Face)GeometricFactors.

       @note Unlike the similarly named protected method UpdateNodes() this
       method does not modify the nodes. */
   void NodesUpdated() { DeleteGeometricFactors(); }
```